qt/scientific_interfaces/Indirect/IndirectFunctionBrowser/MSDFunctionModel.cpp

```diff
@@ -0,0 +1,520 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MSDFunctionModel.h"
+#include "MantidAPI/FunctionFactory.h"
+#include "MantidAPI/IFunction.h"
+#include "MantidAPI/ITableWorkspace.h"
+#include "MantidQtWidgets/Common/FunctionBrowser/FunctionBrowserUtils.h"
+#include <map>
+
+namespace MantidQt {
+namespace CustomInterfaces {
+namespace IDA {
+
+using namespace MantidWidgets;
+using namespace Mantid::API;
+
+namespace {
+std::map<MSDFunctionModel::ParamID, QString> g_paramName{
+    {MSDFunctionModel::ParamID::GAUSSIAN_HEIGHT, "Height"},
+    {MSDFunctionModel::ParamID::GAUSSIAN_MSD, "Msd"},
+    {MSDFunctionModel::ParamID::PETERS_HEIGHT, "Height"},
+    {MSDFunctionModel::ParamID::PETERS_MSD, "Msd"},
+    {MSDFunctionModel::ParamID::PETERS_BETA, "Beta"},
+    {MSDFunctionModel::ParamID::YI_HEIGHT, "Height"},
+    {MSDFunctionModel::ParamID::YI_MSD, "Msd"},
+    {MSDFunctionModel::ParamID::YI_SIGMA, "Sigma"}};
+}
+
+MSDFunctionModel::MSDFunctionModel() {}
+
+void MSDFunctionModel::clearData() {
+  m_fitType.clear();
+  m_model.clear();
+}
+
+void MSDFunctionModel::setFunction(IFunction_sptr fun) {
+  clearData();
+  if (!fun)
+    return;
+  if (fun->nFunctions() == 0) {
+    const auto name = fun->name();
+    if (name == "MsdGauss" || name == "MsdPeters" || name == "MsdYi") {
+      m_fitType = QString::fromStdString(name);
+    } else {
+      throw std::runtime_error("Cannot set function " + name);
+    }
+    m_model.setFunction(fun);
+    return;
+  }
+  bool isFitTypeSet = false;
+  for (size_t i = 0; i < fun->nFunctions(); ++i) {
+    const auto f = fun->getFunction(i);
+    const auto name = f->name();
+    if (name == "MsdGauss" || name == "MsdPeters" || name == "MsdYi") {
+      if (isFitTypeSet) {
+        throw std::runtime_error("Function has wrong structure.");
```

Comment:
Wrong structure in what way? How can a user resolve this since I expect that a user could get their fit function order wrong and run into this

Reply:
I have clarified these messages.